### ISSUE TITLE
Use camelCase for entity types

### ISSUE THREAD

**Issue by egeriis** [redux-json-api/redux-json-api#3]
When storing entities, the type should be camelized to adhere to common JS naming conventions.

This could potentially be a setting, whether to use camelCase or snake_case.

**Comment by antondomratchev** [comment]
I am currently using https://github.com/domchristie/humps in my own implementation.

**Comment by egeriis** [comment]
@antondomratchev How do you feel about that library?

**Comment by morenoh149** [comment]
does it handle circular objects? I tried https://www.npmjs.com/package/camelize but it doesn't detect cycles.

**Comment by egeriis** [comment]
@morenoh149 Circular objects? I'm not sure that is possible in a JSON response? With regards to the _redux-json-api_ state, there are no references between resources.

**Comment by morenoh149** [comment]
ah I misunderstood then. In a project I'm working on we denormalize the JSON response, i.e., make it a circular object client-side.

**Comment by egeriis** [comment]
@morenoh149 Is that to reference related objects?

**Comment by morenoh149** [comment]
yeah. Only downside afaict tell is that the client cannot be isomorphic as you cannot serialize circular javascript objects. It's a design I've been wrestling with at <job> for months.

**Comment by egeriis** [comment]
When the ability comes, I would consider using Proxy objects for this purpose. I've just written a polyfill'ish Proxy to use with _redux-json-api_, it works quite well.

**Comment by evolve2k** [comment]
_I can see everyone points and sense a classic bikeshedding moment potentially arising._ 
Luckily for us JSON API contends that one of it's main benefits is as an "_anti-bikeshedding tool_".

I therefore suggest we follow the JSON API recommendations that "are intended to establish a level of consistency in areas that are beyond the scope of the base JSON API specification."

TLDR: The spec _recommends_  '**caterpillar-case**'.

> ## Naming
> Reference: http://jsonapi.org/recommendations/#naming
>
> The allowed and recommended characters for an URL safe naming of members are defined in the format spec. To also standardize member names, the following (more restrictive) rules are recommended:
> 
> Member names SHOULD start and end with the characters “a-z” (U+0061 to U+007A)
> Member names SHOULD contain only the characters “a-z” (U+0061 to U+007A), “0-9” (U+0030 to U+0039), and the hyphen minus (U+002D HYPHEN-MINUS, “-“) as separator between multiple words.

I think the anti-bikesheading approach then would be to use 'caterpillar-case' (also known as 'kebab-case') as the default. I'd also agree with having an option to change away from the default.

**Comment by egeriis** [comment]
Good point. But caterpillar-case is just not very JavaScript friendly. Imagine having to write this:

`const users = state.api["task-lists"].data[0].attributes["list-name"];`

I much prefer: 
`const users = state.api.taskLists.data[0].attributes.listName;`

**Comment by evolve2k** [comment]
So just to clarify I think we're talking about two naming conventions (that could be the same I suppose). 

One is the naming convention for the end point itself:
localhost:3000/api/v1/food-products
localhost:3000/api/v1/foodProducts

**Comment by morenoh149** [comment]
I think if the project's name has `json-api` in reference to the jsonapi.org spec, you should implement all the recommendations written there. Otherwise why bother right? the point is to conform to single format.

You could always provide alternatives user's can opt into though, like say camelcase.

**Comment by egeriis** [comment]
@morenoh149 The spec doesn't require that attributes are caterpillar-case. But should your server implement all recommendations, there are nothing in _redux-json-api_ that hinders that. 

The motivation for this issue arose to provide a client-side ensurance that all resource types, attributes and relationships are in a particular case. Whether we would provide an ability to specify which casing each implementation would like, is up for discussion. But that's the solution I would lean mostly towards.

But to answer your question, caterpillar-case is _recommendation_ not requirement. I agree that recommendations should be strongly considered, but don't necessarily agree that it is pointless to create a library which only follows requirements.

**Comment by evolve2k** [comment]
> The motivation for this issue arose to provide a client-side ensurance that all resource types, attributes and relationships are in a particular case. Whether we would provide an ability to specify which casing each implementation would like, is up for discussion. But that's the solution I would lean mostly towards.

I would like to suggest that as a next step that we add support for the user to be able to specify the style of key they prefer and once that functionality is available, at that time then set a default to use caterpillar case or dasherized as I've now seen it referred elsewhere.

For some direction on naming conventions, the config of jsonapi-resources has some nice clear naming conventions. I like especially that the json_key_format can be defined separatley from the route_format.

```
  #:underscored_key, :camelized_key, :dasherized_key, or custom
  config.json_key_format = :dasherized_key

  #:underscored_route, :camelized_route, :dasherized_route, or custom
  config.route_format = :dasherized_route
```

I think once configuration like this is possible that you then set the library config to default to dasherized (caterpillar case) to match the official recommendation.

Could be a good inclusion for v2.0

**Comment by egeriis** [comment]
Good idea @evolve2k. If you find the time, feel free to create a PR for v2.0 branch, which includes this feature :)
